crates/uplc/src/machine/cost_model.rs

```diff
@@ -0,0 +1,793 @@
+use crate::builtins::DefaultFunction;
+
+use super::Value;
+
+/// Can be negative
+#[derive(Debug, Clone, PartialEq, Copy)]
+pub struct ExBudget {
+    pub mem: i64,
+    pub cpu: i64,
+}
+
+impl ExBudget {
+    pub fn occurences(&mut self, n: i64) {
+        self.mem *= n;
+        self.cpu *= n;
+    }
+}
+
+impl Default for ExBudget {
+    fn default() -> Self {
+        ExBudget {
+            mem: 14000000,
+            cpu: 10000000000,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct CostModel {
+    pub machine_costs: MachineCosts,
+    pub builtin_costs: BuiltinCosts,
+}
+
+/// There's no entry for Error since we'll be exiting anyway; also, what would
+/// happen if calling 'Error' caused the budget to be exceeded?
+pub struct MachineCosts {
```

Comment:
I'm curious why these are values are assignable, rather than constants? Especially since you'll just use the default?

Reply:
It's nice for experimenting with different cost model values. Eventually we're going to add serde to this so it can be loaded from JSON files.

Reply:
How about having multiple implementations behind an interface? I don't imagine needing to change the costs at runtime.

Reply:
I'm not sure I understand. Let's chat further on discord about this one, atm we've essentially done what they did in Haskell.